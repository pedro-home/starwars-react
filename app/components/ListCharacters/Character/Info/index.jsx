import React from 'react';
import { FlexboxGrid } from 'rsuite';
import PropTypes from 'prop-types';

const Component = ({ type, value }) => {
  return (
    <FlexboxGrid justify="space-between" className="character-info">
      <FlexboxGrid.Item>
        <span className="character-info-type">{type}</span>
      </FlexboxGrid.Item>
      <FlexboxGrid.Item>
        <span className="character-info-value">{value}</span>
      </FlexboxGrid.Item>
    </FlexboxGrid>
  );
};

Component.propTypes = {
  type: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
};

export default Component;
