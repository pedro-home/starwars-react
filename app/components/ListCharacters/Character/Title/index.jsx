import React from 'react';
import PropTypes from 'prop-types';

const Component = ({ text }) => {
  return <h6 className="title">{text}</h6>;
};

Component.propTypes = {
  text: PropTypes.string.isRequired,
};

export default Component;
