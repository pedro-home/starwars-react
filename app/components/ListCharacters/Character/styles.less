.character {
  &-image {
    width: 100%;
    border-bottom: 2px solid @base-color;
  }

  &-entry {
    padding: 0.75rem;
  }
}

@import './Info/styles';
