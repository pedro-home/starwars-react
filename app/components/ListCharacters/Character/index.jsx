import React from 'react';
import { Panel, List } from 'rsuite';
import PropTypes from 'prop-types';
import Info from './Info';
import Title from './Title';

const Component = ({ data, imageUrl }) => {
  return (
    <Panel shaded bodyFill className="character">
      {imageUrl && <img src={imageUrl} className="character-image" />}
      <List>
        {Object.entries(data).map(([key, value]) => (
          <List.Item key={key} className="character-entry">
            {key === 'name' ? <Title text={value} /> : <Info type={key} value={value} />}
          </List.Item>
        ))}
      </List>
    </Panel>
  );
};

Component.propTypes = {
  data: PropTypes.object.isRequired,
  imageUrl: PropTypes.string,
};

export default Component;
