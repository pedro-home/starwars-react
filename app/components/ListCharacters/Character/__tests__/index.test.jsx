import React from 'react';
import { render } from '@testing-library/react';

import Character from '../index';

describe('Given <Character />', () => {
  let component;
  let defaultProps;

  beforeAll(() => {
    defaultProps = { data: { mass: 10, gender: 'Male' } };
  });

  describe('When initializes', () => {
    beforeEach(() => {
      const { container } = render(<Character {...defaultProps} />);
      component = container;
    });

    it('Then should match the snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('And should render data', () => {
      expect(component.querySelectorAll('.character-entry').length).toEqual(Object.keys(defaultProps.data).length);
    });
  });

  describe('And has a name in data', () => {
    let name;
    beforeEach(() => {
      name = 'joker';
      const { container } = render(<Character {...defaultProps} data={{ ...defaultProps.data, name }} />);
      component = container;
    });

    it('Then should match the snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('Then should render a title', () => {
      const elems = component.querySelectorAll('.character-entry .title');
      expect(elems.length).toEqual(1);
      expect(elems[0].innerHTML).toEqual(name);
    });
  });

  describe('And has an image', () => {
    let imageUrl;
    beforeEach(() => {
      imageUrl = 'testing.png';
      const { container } = render(<Character {...defaultProps} imageUrl={imageUrl} />);
      component = container;
    });

    it('Then should match the snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('Then should render the character image', () => {
      expect(component.querySelector('.character-image').getAttribute('src')).toEqual(imageUrl);
    });
  });
});
