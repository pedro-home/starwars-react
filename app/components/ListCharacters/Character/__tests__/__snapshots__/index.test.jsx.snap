// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <Character /> And has a name in data Then should match the snapshot 1`] = `
<div>
  <div
    class="character rs-panel rs-panel-default rs-panel-shaded"
  >
    <div
      class="rs-panel-body rs-panel-body-fill"
    >
      <div
        class="rs-list"
      >
        <div
          class="rs-list-item character-entry rs-list-item-md"
          manager="[object Object]"
        >
          <div
            class="rs-list-item-content"
          >
            <div
              class="rs-flex-box-grid character-info rs-flex-box-grid-top rs-flex-box-grid-space-between"
            >
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-type"
                >
                  mass
                </span>
              </div>
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-value"
                >
                  10
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rs-list-item character-entry rs-list-item-md"
          manager="[object Object]"
        >
          <div
            class="rs-list-item-content"
          >
            <div
              class="rs-flex-box-grid character-info rs-flex-box-grid-top rs-flex-box-grid-space-between"
            >
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-type"
                >
                  gender
                </span>
              </div>
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-value"
                >
                  Male
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rs-list-item character-entry rs-list-item-md"
          manager="[object Object]"
        >
          <div
            class="rs-list-item-content"
          >
            <h6
              class="title"
            >
              joker
            </h6>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Given <Character /> And has an image Then should match the snapshot 1`] = `
<div>
  <div
    class="character rs-panel rs-panel-default rs-panel-shaded"
  >
    <div
      class="rs-panel-body rs-panel-body-fill"
    >
      <img
        class="character-image"
        src="testing.png"
      />
      <div
        class="rs-list"
      >
        <div
          class="rs-list-item character-entry rs-list-item-md"
          manager="[object Object]"
        >
          <div
            class="rs-list-item-content"
          >
            <div
              class="rs-flex-box-grid character-info rs-flex-box-grid-top rs-flex-box-grid-space-between"
            >
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-type"
                >
                  mass
                </span>
              </div>
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-value"
                >
                  10
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rs-list-item character-entry rs-list-item-md"
          manager="[object Object]"
        >
          <div
            class="rs-list-item-content"
          >
            <div
              class="rs-flex-box-grid character-info rs-flex-box-grid-top rs-flex-box-grid-space-between"
            >
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-type"
                >
                  gender
                </span>
              </div>
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-value"
                >
                  Male
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Given <Character /> When initializes Then should match the snapshot 1`] = `
<div>
  <div
    class="character rs-panel rs-panel-default rs-panel-shaded"
  >
    <div
      class="rs-panel-body rs-panel-body-fill"
    >
      <div
        class="rs-list"
      >
        <div
          class="rs-list-item character-entry rs-list-item-md"
          manager="[object Object]"
        >
          <div
            class="rs-list-item-content"
          >
            <div
              class="rs-flex-box-grid character-info rs-flex-box-grid-top rs-flex-box-grid-space-between"
            >
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-type"
                >
                  mass
                </span>
              </div>
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-value"
                >
                  10
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="rs-list-item character-entry rs-list-item-md"
          manager="[object Object]"
        >
          <div
            class="rs-list-item-content"
          >
            <div
              class="rs-flex-box-grid character-info rs-flex-box-grid-top rs-flex-box-grid-space-between"
            >
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-type"
                >
                  gender
                </span>
              </div>
              <div
                class="rs-flex-box-grid-item rs-flex-box-grid-item-0"
              >
                <span
                  class="character-info-value"
                >
                  Male
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
